test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt

```diff
@@ -1,200 +1,206 @@
-	$noteID = fopen('php://memory', 'a');
-	$data = '0';
-	$query = "UPDATE PDO_AllTypes SET [VarbinaryCol]=:Name WHERE [BitCol]=:value";
-	$stmt = $db->prepare($query,array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
-	fwrite($noteID, '00');
-	rewind($noteID);
-	$stmt->bindParam(':Name',$noteID,PDO::PARAM_LOB, 0, PDO::SQLSRV_ENCODING_BINARY);
-	$stmt->bindParam(':value',$data);
-	$result = $stmt->execute();
-	
-	$select = "SELECT * FROM PDO_AllTypes WHERE [BitCol]=:value";
-	$stmt = $db->prepare($select,array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
-	$stmt->bindParam(':value',$data);
-	$stmt->execute();
-	$result = $stmt->fetchColumn(26);	// Retrieve VarbinaryCol
-	print("$result\n");
+    // Binding LOB with buffered queries activated.
+    $noteID = fopen('php://memory', 'a');
+    $data = '0';
+    $query = "UPDATE $tbname SET [VarbinaryCol]=:Name WHERE [BitCol]=:value";
+    $stmt = $db->prepare($query, array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
+    fwrite($noteID, '00');
+    rewind($noteID);
+    $stmt->bindParam(':Name', $noteID, PDO::PARAM_LOB, 0, PDO::SQLSRV_ENCODING_BINARY);
+    $stmt->bindParam(':value', $data);
+    $result = $stmt->execute();
+
+    $select = "SELECT * FROM $tbname WHERE [BitCol]=:value";
+    $stmt = $db->prepare($select, array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
+    $stmt->bindParam(':value', $data);
+    $stmt->execute();
+    $result = $stmt->fetchColumn(26);    // Retrieve VarbinaryCol
+    unset($stmt);
+    print("$result\n");
 }
 
-function bindParam_STR($db)
+function bindPARAM_STR($db, $tbname)
 {
-	// Binding STR with buffered queries activated.
-	$noteID = fopen('php://memory', 'a');
-	$data = 'STRINGCOL2';
-	$query = "UPDATE PDO_AllTypes SET [BitCol]=:Name WHERE [CharCol]=:value";
-	$stmt = $db->prepare($query,array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
-	fwrite($noteID, '1');
-	rewind($noteID);
-	$stmt->bindParam(':Name',$noteID,PDO::PARAM_STR, 0, PDO::SQLSRV_ENCODING_BINARY);
-	$stmt->bindParam(':value',$data);
-	$stmt->execute();
-	
-	$select = "SELECT * FROM PDO_AllTypes WHERE [CharCol]=:value";
-	$stmt = $db->prepare($select,array(PDO::ATTR_CURSOR=>PDO::CURSOR_SCROLL, PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE=>PDO::SQLSRV_CURSOR_BUFFERED));
-	$stmt->bindParam(':value',$data);
-	$stmt->execute();
-	$result = $stmt->fetchColumn(2);	// Retrieve BitCol
-	$result = str_replace("\0","",$result);
-	print("$result\n");
-	
+    // Binding STR with buffered queries activated.
+    // fopen returns a resource, if using PDO::PARAM_STR, the PDO side converts it to a zend_string "Resource id #*"
+    // thus a resource is not meant to bind parameter using PDO::PARAM_STR
+
```

Comment:
Does that mean the design of the original test is flawed? So even with PDO::SQLSRV_ENCODING_BINARY specified doesn't work as expected?

Reply:
Yes the original test is flawed. With PDO::PARAM_STR, the PDO side already convert it to "Resource id #" when it gets to our driver. Setting the encoding the binary only changes how we pass in "Resource id #" to the ODBC driver.